### PR TITLE
Split Docker Build to accept build args to optionally include swift support

### PR DESCRIPTION
**Pull request by nlaurance-pyie**
added dockerignore for slimer build context
parametrized dockerfile

making swift support optional

nil/bandersnatch         3.9                 9f730c999230        2 minutes ago       156MB
nil/bandersnatch         3.9-swift           91d50bd7103d        4 minutes ago       280MB

hopefully addressing https://github.com/pypa/bandersnatch/issues/740 also

### CODE REVIEW

File: Dockerfile
Comment (cooperlees) [inline Dockerfile:1]: Sweet - 1 spot is way better!

**Review by cooperlees — CHANGES_REQUESTED** [review]
Looks good, but I think it's going to break automation.

2 questions:
1) How des this remove the setuptools dep? We use import from it
e.g. filter.py + storage.py import pkg_resources
- https://github.com/pypa/bandersnatch/blob/master/src/bandersnatch/filter.py#L7
- https://github.com/pypa/bandersnatch/blob/master/src/bandersnatch/storage.py#L25
2) I feel this will need to pass this arg to the docker building GitHub action:
https://github.com/pypa/bandersnatch/blob/master/.github/workflows/docker_upload.yml
- Seems you just need to add `buildargs: PY_VERSION=3.9`

**Comment by nlaurance-pyie** [comment]
1/ poor phrasing, sorry. it's just out from requirements and setup. Still needed tho. its presence triggered warning

**Comment by nlaurance-pyie** [comment]
`!	refs/heads/docker_py_version:refs/heads/docker_py_version	[remote rejected] (refusing to allow an OAuth App to create or update workflow `.github/workflows/docker_upload.yml` without `workflow` scope)`


```
name: bandersnatch_docker_upload

on:
  push:
    branches:
      - master

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@master
      - name: Publish to Docker Registry
        uses: elgohr/Publish-Docker-Github-Action@master
        env:
          PY_VERSION: 3.9
        with:
          name: pypa/bandersnatch
          username: ${{ secrets.DOCKER_USERNAME }}
          password: ${{ secrets.DOCKER_PASSWORD }}
          snapshot: true
          buildargs: PY_VERSION
          tags: "swift-latest, swift-3.9, swift-3"
      - name: Publish Swift variation to Docker Registry
        uses: elgohr/Publish-Docker-Github-Action@master
        env:
          PY_VERSION: 3.9
          WITH_SWIFT: yes
        with:
          name: pypa/bandersnatch
          username: ${{ secrets.DOCKER_USERNAME }}
          password: ${{ secrets.DOCKER_PASSWORD }}
          snapshot: true
          buildargs: PY_VERSION, WITH_SWIFT
          tags: "latest, 3.9, 3"

```

**Comment by cooperlees** [comment]
> 1/ poor phrasing, sorry. it's just out from requirements and setup. Still needed tho. its presence triggered warning

What warning? Seems weird to remove something we know is a dep. If you can't work it out can you leave it and open an issue and I'll try fix it.

> ! refs/heads/docker_py_version:refs/heads/docker_py_version [remote rejected] (refusing to allow an OAuth App to create or update workflow .github/workflows/docker_upload.ymlwithoutworkflow scope)

What's this? Do I need to do something here?

**Comment by nlaurance-pyie** [comment]
1/ issue is similar to https://github.com/pypa/pipenv/issues/1417 (setuptools shouldn't be a dependency since you cannot install dependencies without it in the first place)
now, I don't mind to put it back

the other matter is it seems I cannot push any changes in the .github repo. And you would need to evaluate my changes yourself, I'm afraid. It makes the image without swift support the standard one, and also build a variation with the optional swift support.

**Comment by cooperlees** [comment]
Really, `.github` should just checkout like any other directory. Worst comes to worst I can look at merging your changes, but there is no protection to the `.github` directory that I know of.

**Comment by nlaurance-pyie** [comment]
.github is checked out alright, but I can't push .
it seems a security feature.

Also, that Dockerfile doesn't build on raspberry pi ;(

**Comment by cooperlees** [comment]
TIL. Ok, well I can fix that post merging this.

Raspberry Pi has never been a target platform. File an issue with the errors + a repro and I or someone else will see if we can fix it.

**Comment by cooperlees** [comment]
Please fix lint:
Fix End of Files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook
Fixing MAINTAINERS.md
Fixing .dockerignore

**Comment by cooperlees** [comment]
I'm looking to release this week so if you could fix the lints + put setuptools back so we can see what version we test with and what version each release / docker container uses that would be grand.

**Review by cooperlees — APPROVED** [review]

**Comment by cooperlees** [comment]
Thanks. Will merge soon when I have time to fix the docker image building etc. right after merging.

**Comment by cooperlees** [comment]
This still build and uploaded a docker image:
https://github.com/pypa/bandersnatch/runs/1471070904

Will try a PR with your suggested changes tho.

**Pull request merged**
